presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/EquivalenceClassProperty.java

```diff
@@ -0,0 +1,311 @@
+            members.forEach(member -> updateInternal(head, member));
+        });
+    }
+
+    /**
+     * Updates this equivalence class property with "variable reference = variable reference" or
+     * "variable reference = constant reference" conjuncts applied by the provided predicate.
+     * Returns true if any equivalence class heads changed, and false otherwise.
+     *
+     * @param predicate
+     * @return
+     */
+    public boolean update(RowExpression predicate)
+    {
+        requireNonNull(predicate, "predicate is null");
+        updated = false;
+        //TODO tunnel through CAST functions?
+        extractConjuncts(predicate).stream()
+                .filter(CallExpression.class::isInstance)
+                .map(CallExpression.class::cast)
+                .filter(e -> isVariableEqualVariableOrConstant(e))
+                .forEach(e -> updateInternal(e.getArguments().get(0), e.getArguments().get(1)));
+        return updated;
+    }
+
+    private boolean isVariableEqualVariableOrConstant(RowExpression expression)
+    {
+        if (expression instanceof CallExpression
+                && functionResolution.isEqualFunction(((CallExpression) expression).getFunctionHandle())
+                && ((CallExpression) expression).getArguments().size() == 2) {
+            RowExpression e1 = ((CallExpression) expression).getArguments().get(0);
+            RowExpression e2 = ((CallExpression) expression).getArguments().get(1);
+
+            if ((e1 instanceof VariableReferenceExpression && (e2 instanceof VariableReferenceExpression || e2 instanceof ConstantExpression))
+                    || ((e2 instanceof VariableReferenceExpression) && (e1 instanceof VariableReferenceExpression || e1 instanceof ConstantExpression))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void update(RowExpression firstExpression, RowExpression secondExpression)
+    {
+        updateInternal(firstExpression, secondExpression);
+    }
+
+    /**
+     * Updates this equivalence class property with pairs of variable or column references deemed
+     * equivalent via the application of predicates.
+     * Side effect is that it sets updated instance variable true if any equivalence class heads changed.
+     * This can be used to optimize methods that maintain alignment with equivalence classes.
+     *
+     * @param firstExpression
+     * @param secondExpression
+     */
+    private void updateInternal(RowExpression firstExpression, RowExpression secondExpression)
+    {
+        RowExpression head1 = getEquivalenceClassHead(firstExpression);
+        RowExpression head2 = getEquivalenceClassHead(secondExpression);
+
+        //already in same equivalence class, nothing to do
+        //note that we do not check head1.equal(head2) so that two different variable reference objects
+        //referencing the same reference are both added to the equivalence class
```

Comment:
@simmend can you explain why we want this to be true?  (and it seems like when we use it, it's always as part of a set, or inside a map, so the distinction isn't preserved).